### PR TITLE
Different API Change.

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

**Comment by fulmicoton** [comment]
Do not merge! This is a work in progress.

Experiment on changing the API.

**Comment by poljar** [comment]
The changes look OK. 

Now I did intentionally leave types in the bindings and avoided generic dictionaries or tuples since I prefer specialized  types/classes to generic dicts and a stringly typed API but I am also aware that most people prefer passing just a dicts with keys/values, especially if they are working with json.

So conclusion is, I'm fine with the changes and won't press for more types in the API.

File: Cargo.toml
Comment (bloodbare) [inline Cargo.toml]: Could we get it back to global? I can't compile and test it on my local requirements dependency

File: src/index.rs
Comment (bloodbare) [inline src/index.rs]:
```
no method named `add_named_document` found for type `tantivy::indexer::index_writer::IndexWriter` in the current scope
```

File: src/index.rs
Comment (bloodbare) [inline src/index.rs]:
```
no method named `add_json` found for type `tantivy::indexer::index_writer::IndexWriter` in the current scope
```

**Comment by fulmicoton** [comment]
@poljar 

Let me see if we can get the best of two worlds...

What about we reintroduce the `Document` type, make it possible to build it from a dict, and 
build the document type using string as field names (still no Field type)

**Comment by fulmicoton** [comment]
@polar @bloodbare 

Ok can you guys review the code please?
The CR is  quite heavy sorry...

There was a complicated bug in the way dates were converted. Here there is still a problem,
but at least it works correctly for dates that do not have any timezone.

Not defining any default fields is ok now. By default, tantivy will search all indexed fields.

**Comment by bloodbare** [comment]
I've been away also, testing and reviewing next days!

File: src/index.rs
Comment (bloodbare) [inline src/index.rs]:
```
           Fresh tantivy v0.11.0 (https://github.com/tantivy-search/tantivy#ea56160c)
         Running `rustc --edition=2018 --crate-name pyo3 /Users/ramon/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/lib.rs --color never --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="extension-module"' --cfg 'feature="python3"' -C metadata=67ff347bbf359588 -C extra-filename=-67ff347bbf359588 --out-dir /Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps -L dependency=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps --extern inventory=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libinventory-2e5c33c844a8887c.rlib --extern libc=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/liblibc-39401ca11bee5cf9.rlib --extern mashup=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libmashup-2abbf69123ea527e.rlib --extern num_traits=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libnum_traits-a00981b0317fdbac.rlib --extern pyo3cls=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libpyo3cls-15f681bf5e02d088.dylib --extern spin=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libspin-fcbf37cb1a18d91d.rlib --cap-lints allow --cfg Py_3_5 --cfg Py_3_6 --cfg Py_3_7 --cfg Py_3 --cfg 'py_sys_config="WITH_THREAD"'`
       Compiling tantivy-py v0.10.1 (/Users/ramon/stashify/backend/processing/index/src/tantivy)
         Running `rustc --edition=2018 --crate-name tantivy src/lib.rs --color never --crate-type cdylib --emit=dep-info,link -C debuginfo=2 --crate-type cdylib -C link-arg=-undefined -C link-arg=dynamic_lookup --cfg 'feature="pyo3"' -C metadata=72a5d54f72dad5ee --out-dir /Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps -C incremental=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/incremental -L dependency=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps --extern chrono=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libchrono-7d3f9e16397c328a.rlib --extern itertools=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libitertools-32bcac7bdb1e0f6d.rlib --extern pyo3=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libpyo3-67ff347bbf359588.rlib --extern tantivy=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/deps/libtantivy-ad34e0a1506b0ecc.rlib -L native=/Users/ramon/stashify/backend/processing/index/src/tantivy/target/debug/build/backtrace-sys-a6dbac9fe5ce7459/out`
    error[E0599]: no method named `convert_named_doc` found for type `tantivy::schema::schema::Schema` in the current scope
      --> src/index.rs:39:31
       |
    39 |         let doc = self.schema.convert_named_doc(named_doc).map_err(to_pyerr)?;
       |                               ^^^^^^^^^^^^^^^^^

    error: aborting due to previous error

    For more information about this error, try `rustc --explain E0599`.
    error: Could not compile `tantivy-py`.
```

**Review by bloodbare — COMMENTED** [review]
Great work! I'm not able to compile it right now!

**Review by bloodbare — COMMENTED** [review]
Would be great to map also the tokenizer API, so its possible to define ngram tokenizers

**Comment by fulmicoton** [comment]
@bloodbare I froze tantivy's revision. Can you tell me if that fixed the build for you?

**Comment by bloodbare** [comment]
> @bloodbare I froze tantivy's revision. Can you tell me if that fixed the build for you?

Yups, thanks! Reviewing

File: src/document.rs
Comment (bloodbare) [inline src/document.rs]: Facet values fail on casting if they are not already Facet objects

Comment (fulmicoton) [reply src/document.rs]: To be clear, you mean we need this to work if they are strings too correct?

File: tantivy/__init__.py
Comment (poljar) [inline tantivy/__init__.py]: Is there a reason why we are removing these?

Many unix tools expect them to be there, e.g:
```bash
$ wc -l tantivy/__init__.py 
0 tantivy/__init__.py
```

File: tests/tantivy_test.py
Comment (poljar) [inline tests/tantivy_test.py]: Another file without a newline.

**Review by poljar — CHANGES_REQUESTED** [review]
Overall happy with the changes, great work.

After the small nitpicks are resolved we can go ahead and merge this.

File: src/document.rs
Comment (bloodbare) [reply src/document.rs]: yups, its easier from API point of view to not needing to cast as facet objects, imho.

**Comment by bloodbare** [comment]
Two missing API to complete the circle are the `delete_term` on indexerwriter and the `Term::from_field_text` object mapping. I don't know if this should be a new issue.

File: src/document.rs
Comment (fulmicoton) [reply src/document.rs]: addressed

File: src/index.rs
Comment (fulmicoton) [reply src/index.rs]: This is not exactly an accessor. This picks a searcher from a pool. It has some side effect that should be in the user radar. I'll add some doc to explain that.

File: tantivy/__init__.py
Comment (fulmicoton) [reply tantivy/__init__.py]: no reason I am just making a mistake. I did not know about `wc -l` not counting that as a line though. 
I like a trailing \n for `cat`

File: src/document.rs
Comment (fulmicoton) [reply src/document.rs]: This seem like a good idea. Can you open an issue and double check with @poljar we want this?

File: src/document.rs
Comment (fulmicoton) [reply src/document.rs]: I am not fond of `__len__` or len here as it is quite ambiguous as field can be multivalued.

Would that mean `num_fields` or `num_field_values` ?

Comment (fulmicoton) [reply src/document.rs]: If you want this we can keep this method and also have `__len__` call it I guess.

File: src/document.rs
Comment (fulmicoton) [reply src/document.rs]: Agree on opening separate issues.

**Comment by fulmicoton** [comment]
> circle are the `delete_term` on indexerwriter and the `Term::from_field_text` object m

Can you open a separate issue?

**Comment by fulmicoton** [comment]
@poljar I think I integrated your comments. Thanks for the review. I did not implement `__len__` nor `__bool__` . We can do that in a different issue I think.

**Comment by poljar** [comment]
Can we just remove the revision for tantivy in the Cargo.toml before we merge this. I would like to keep it working with the latest master. I tested it and the PR works with latest master of tantivy over here.

Otherwise this looks good to merge. Thanks.

**Review by poljar — APPROVED** [review]

**Comment by fulmicoton** [comment]
> Can we just remove the revision for tantivy in the Cargo.toml before we merge this. I would like to keep it working with the latest master. I tested it and the PR works with latest master of tantivy over here.

I think this is a very bad idea but you are the boss.
Removing the revision and merging.

**Comment by poljar** [comment]
Well we will soon™ depend on the next version of Tantivy, until that happens it will be useful to see if stuff breaks ASAP.

**Pull request merged**
